[SPARK-14238][ML][MLLIB][PYSPARK] Add binary toggle Param to PySpark HashingTF in ML & MLlib

yongtang opened the pull request:
## What changes were proposed in this pull request?

This fix tries to add binary toggle Param to PySpark HashingTF in ML & MLlib. If this toggle is set, then all non-zero counts will be set to 1.

Note: This fix (SPARK-14238) is extended from SPARK-13963 where Scala implementation was done.
## How was this patch tested?

This fix adds two tests to cover the code changes. One for HashingTF in PySpark's ML and one for HashingTF in PySpark's MLLib.

yongtang committed:
[SPARK-14238][ML][MLLIB][PYSPARK] Add binary toggle Param to PySpark HashingTF in ML & MLlib

This fix tries to add binary toggle Param to PySpark HashingTF in ML & MLlib.
If this toggle is set, then all non-zero counts will be set to 1.

This fix adds two tests to cover the code changes. One for HashingTF in PySpark's ML
and one for HashingTF in PySpark's MLLib.

AmplabJenkins commented:
Can one of the admins verify this patch?

MLnick commented:
ok to test

SparkQA commented:
**[Test build #54623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54623/consoleFull)** for PR 12079 at commit [`e58d1a2`](https://github.com/apache/spark/commit/e58d1a279aaded9045c9e7a7a161500163b81fd6).

SparkQA commented:
**[Test build #54623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54623/consoleFull)** for PR 12079 at commit [`e58d1a2`](https://github.com/apache/spark/commit/e58d1a279aaded9045c9e7a7a161500163b81fd6).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54623/
Test FAILed.

yongtang committed:
Fix PEP8 errors.

SparkQA commented:
**[Test build #54631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54631/consoleFull)** for PR 12079 at commit [`1e24a68`](https://github.com/apache/spark/commit/1e24a68806e616996b3e465f38bbbd3525a18b11).

SparkQA commented:
**[Test build #54631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54631/consoleFull)** for PR 12079 at commit [`1e24a68`](https://github.com/apache/spark/commit/1e24a68806e616996b3e465f38bbbd3525a18b11).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54631/
Test PASSed.

yanboliang commented on python/pyspark/ml/feature.py:
```diff
@@ -512,6 +512,16 @@ class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, Java
+    (default = False)
+    """
+    binary = Param(Params._dummy(), "binary",
+                   "Binary toggle to control term frequency counts",
```
We should keep the doc of the Param consistent with Scala.

yanboliang commented on python/pyspark/ml/feature.py:
```diff
@@ -512,6 +512,16 @@ class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, Java
     """
 
+    """
+    Binary toggle to control term frequency counts.
```
The comment doc for ```binary``` is unnecessary because it will not exist in the generated Python API doc.

yanboliang commented on python/pyspark/ml/feature.py:
```diff
@@ -520,6 +530,7 @@ def __init__(self, numFeatures=1 << 18, inputCol=None, outputCol=None):
         super(HashingTF, self).__init__()
         self._java_obj = self._new_java_obj("org.apache.spark.ml.feature.HashingTF", self.uid)
         self._setDefault(numFeatures=1 << 18)
+        self._setDefault(binary=False)
```
```self._setDefault(numFeatures=1 << 18, binary=False)```

yanboliang commented on python/pyspark/ml/feature.py:
```diff
@@ -512,6 +512,16 @@ class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, Java
+                   typeConverter=TypeConverters.toBoolean)
+
     @keyword_only
     def __init__(self, numFeatures=1 << 18, inputCol=None, outputCol=None):
```
```binary``` should be arguments of ```__init__```.

yongtang replied:
Thanks @yanboliang just updated the pull request with issues addressed. Let me know if there are  any other issues.

yongtang committed:
Update comment blocks and args according to @yanboliang's comment.

SparkQA commented:
**[Test build #54642 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54642/consoleFull)** for PR 12079 at commit [`a71f59b`](https://github.com/apache/spark/commit/a71f59b567c998c19df362123e587fef444e6db7).

SparkQA commented:
**[Test build #54642 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54642/consoleFull)** for PR 12079 at commit [`a71f59b`](https://github.com/apache/spark/commit/a71f59b567c998c19df362123e587fef444e6db7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54642/
Test PASSed.

yongtang commented:
Rebased to fix conflicts.

SparkQA commented:
**[Test build #55524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55524/consoleFull)** for PR 12079 at commit [`829c87e`](https://github.com/apache/spark/commit/829c87e453b7082edd3d5b9d042b9fd849f060b8).

SparkQA commented:
**[Test build #55524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55524/consoleFull)** for PR 12079 at commit [`829c87e`](https://github.com/apache/spark/commit/829c87e453b7082edd3d5b9d042b9fd849f060b8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55524/
Test PASSed.

holdenk commented:
One minor note:Often we want to go with Scala first then Python, but in either direction if we are only doing one at a time it can be good practice to create either a follow up JIRA or a subtask on the existing JIRA to also expose the implementation in the other language.

holdenk commented on python/pyspark/ml/feature.py:
```diff
@@ -512,14 +512,19 @@ class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, Java
     .. versionadded:: 1.3.0
     """
 
+    binary = Param(Params._dummy(), "binary", "If true, all non zero counts are set to 1. " +
```
We probably want to mention the default value here (namely false).

yongtang replied:
Thanks @holdenk this issue has been addressed.

holdenk replied:
Great! Looking at the incoming PRs it seems there is a second PR also adding a binary feature to another model - it might make sense to move this to a shared param instead of having it be per-model (although it will require coordination with the other PR timing wise).

MLnick replied:
See https://github.com/apache/spark/pull/12308#issuecomment-208773214

MLnick replied:
`if true` -> `if True`

BryanCutler commented on python/pyspark/mllib/feature.py:
```diff
@@ -379,6 +379,17 @@ class HashingTF(object):
+        """
+        If true, term frequency vector will be binary such that non-zero
+        term counts will be set to 1
+        (default: false)
```
minor: false -> False

yongtang replied:
Thanks @BryanCutler this issue has been corrected.

yongtang committed:
Minor fix in comment based on feedbacks from @holdenk and @BryanCutler.

SparkQA commented:
**[Test build #55587 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55587/consoleFull)** for PR 12079 at commit [`9c2b4ab`](https://github.com/apache/spark/commit/9c2b4ab64fbd0aa267ff1ee0b9b353a99346d05c).

yongtang commented:
@holdenk The Scala implementation has ben completed in SPARK-13963. I updated the description of this pull request to show the linkage between this issue (SPARK-14238) and SPARK-13963.

SparkQA commented:
**[Test build #55587 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55587/consoleFull)** for PR 12079 at commit [`9c2b4ab`](https://github.com/apache/spark/commit/9c2b4ab64fbd0aa267ff1ee0b9b353a99346d05c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55587/
Test PASSed.

MLnick commented on python/pyspark/ml/feature.py:
```diff
@@ -512,14 +512,19 @@ class HashingTF(JavaTransformer, HasInputCol, HasOutputCol, HasNumFeatures, Java
 
+    binary = Param(Params._dummy(), "binary", "If true, all non zero counts are set to 1. " +
+                   "This is useful for discrete probabilistic models that model binary events " +
+                   "rather than integer counts. (default: False)",
```
The style seems to be `. Default False` rather than `. (default: False)`. @BryanCutler @holdenk thoughts?

Though I must say I'd prefer `(default: X).` across the board myself.

MLnick commented on python/pyspark/ml/tests.py:
```diff
@@ -831,6 +831,25 @@ def test_logistic_regression_summary(self):
+        hashingTF.setInputCol("words").setOutputCol("features").setNumFeatures(n).setBinary(True)
+        output = hashingTF.transform(df)
+        features = output.select("features").first().features.toArray()
+        expected = Vectors.sparse(100, {(ord("a") % n): 1.0,
```
`100` -> `n`

MLnick commented on python/pyspark/mllib/feature.py:
```diff
@@ -379,6 +379,17 @@ class HashingTF(object):
+    @since("2.0.0")
+    def setBinary(self, value):
+        """
+        If true, term frequency vector will be binary such that non-zero
```
`if true` -> `if True`

MLnick commented:
A few minor comments, otherwise LGTM. 

@holdenk @BryanCutler we could merge this and #12308, and then update the param to be shared (if we can do the different doc thing?).

yongtang committed:
Several comment area fixes based on comments from @MLnick.

SparkQA commented:
**[Test build #55608 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55608/consoleFull)** for PR 12079 at commit [`551cc6e`](https://github.com/apache/spark/commit/551cc6ee1b4fdb5cff59d4ef998ca0a15777c7e7).

yongtang commented:
Thanks @MLnick I just updated the pull request to address several minor issues. With respect to `. Default False` vs `. (default: False)`, I changed it to `. Default False` for now. But if you want to see `(default: X)` I can change it (including the rest of the file) to it as well.

SparkQA commented:
**[Test build #55608 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55608/consoleFull)** for PR 12079 at commit [`551cc6e`](https://github.com/apache/spark/commit/551cc6ee1b4fdb5cff59d4ef998ca0a15777c7e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55608/
Test PASSed.

BryanCutler commented:
> @holdenk @BryanCutler we could merge this and #12308, and then update the param to be shared (if we can do the different doc thing?).

I think that will be better and maybe then we can change the param to be shared on both the Scala and Python side.

holdenk commented:
@BryanCutler / @yongtang That sounds reasonable :)

MLnick commented:
As per @jkbradley's https://github.com/apache/spark/pull/12308#issuecomment-209039855, let's keep them separate params.

MLnick commented:
jenkins retest this please

SparkQA commented:
**[Test build #55839 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55839/consoleFull)** for PR 12079 at commit [`551cc6e`](https://github.com/apache/spark/commit/551cc6ee1b4fdb5cff59d4ef998ca0a15777c7e7).

SparkQA commented:
**[Test build #55839 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55839/consoleFull)** for PR 12079 at commit [`551cc6e`](https://github.com/apache/spark/commit/551cc6ee1b4fdb5cff59d4ef998ca0a15777c7e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55839/
Test PASSed.

MLnick commented:
LGTM. Merged to master.
